internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
 	}
 
-	if txmp.recheckCursor == nil {
-		txmp.logger.Debug("finished rechecking transactions")
-
-		if txmp.Size() > 0 {
-			txmp.notifyTxsAvailable()
-		}
+	if checkTxRes.CheckTx.Code == abci.CodeTypeOK && err == nil {
+		wtx.SetPriority(checkTxRes.CheckTx.Priority)
+		return // N.B. Size of mempool did not change
 	}
 
+	txmp.logger.Debug(
+		"existing transaction no longer valid; failed re-CheckTx callback",
+		"priority", wtx.Priority(),
+		"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+		"err", err,
+		"code", checkTxRes.CheckTx.Code,
+	)
+	txmp.removeTxByElement(elt)
+	if !txmp.config.KeepInvalidTxsInCache {
+		txmp.cache.Remove(wtx.tx)
+	}
 	txmp.metrics.Size.Set(float64(txmp.Size()))
 }
 
-// updateReCheckTxs updates the recheck cursors by using the gossipIndex. For
-// each transaction, it executes CheckTxAsync. The global callback defined on
-// the proxyAppConn will be executed for each transaction after CheckTx is
-// executed.
+// recheckTransactions initiates re-CheckTx ABCI calls for all the transactions
+// currently in the mempool. It reports the number of recheck calls that were
+// successfully initiated.
 //
-// NOTE:
-// - The caller must have a write-lock when executing updateReCheckTxs.
-func (txmp *TxMempool) updateReCheckTxs() {
+// Precondition: The mempool is not empty.
+// The caller must hold txmp.mtx exclusively.
+func (txmp *TxMempool) recheckTransactions() {
 	if txmp.Size() == 0 {
-		panic("attempted to update re-CheckTx txs when mempool is empty")
+		panic("mempool: cannot run recheck on an empty mempool")
 	}
+	txmp.logger.Debug(
+		"executing re-CheckTx for all remaining transactions",
+		"num_txs", txmp.Size(),
+		"height", txmp.height,
+	)
+	// N.B.: We have to issue the calls outside the lock. In a local client,
+	// even an "async" call invokes its callback immediately which will make the
+	// callback deadlock trying to acquire the same lock. This isn't a problem
+	// with out-of-process calls, but this has to work for both.
+	txmp.mtx.Unlock()
+	defer txmp.mtx.Lock()
 
-	txmp.recheckCursor = txmp.gossipIndex.Front()
-	txmp.recheckEnd = txmp.gossipIndex.Back()
-	ctx := context.Background()
-
-	for e := txmp.gossipIndex.Front(); e != nil; e = e.Next() {
+	ctx := context.TODO()
+	atomic.StoreInt64(&txmp.txRecheck, int64(txmp.txs.Len()))
+	for e := txmp.txs.Front(); e != nil; e = e.Next() {
 		wtx := e.Value.(*WrappedTx)
 
-		// Only execute CheckTx if the transaction is not marked as removed which
-		// could happen if the transaction was evicted.
-		if !txmp.txStore.IsTxRemoved(wtx.hash) {
-			_, err := txmp.proxyAppConn.CheckTxAsync(ctx, abci.RequestCheckTx{
-				Tx:   wtx.tx,
-				Type: abci.CheckTxType_Recheck,
-			})
-			if err != nil {
-				// no need in retrying since the tx will be rechecked after the next block
-				txmp.logger.Error("failed to execute CheckTx during rechecking", "err", err)
-			}
+		_, err := txmp.proxyAppConn.CheckTxAsync(ctx, abci.RequestCheckTx{
```

Comment:
Can you add a comment here perhaps that this function will call `recheckTxCallback` which is responsible for removing the transactions else for the untrained reader it might not be clear that that is happening